### PR TITLE
Allows to override themes from the app

### PR DESCRIPTION
**Pull request by phillaf**
Folder structure is the following
src/Template/MyTheme/Articles/view.ctp

and plugins overrided like this
src/Template/MyTheme/Plugin/MyPlugin/Users/view.ctp

see #7241

### CODE REVIEW

**Comment by Phillaf** [comment]
I'll update the tests :/

File: tests/TestCase/View/ViewTest.php [hunk 1]
```diff
@@ -492,6 +494,8 @@ public function testPathThemedPluginGeneration()
             TEST_APP . 'TestApp' . DS . 'Template' . DS,
             CAKE . 'Template' . DS,
         ];
+        debug($paths);
```
Comment (antograssiot) [inline tests/TestCase/View/ViewTest.php]: @Phillaf you probably didn't want to leave those lines

Comment (Phillaf) [inline tests/TestCase/View/ViewTest.php]: :flushed:

**Comment by Phillaf** [comment]
There are still some issues with Emails. I'll try to commit the fix today.

**Comment by markstory** [comment]
Looks good to me.

**Comment by lorenzo** [comment]
The list of fallback paths is growing strong :S

**Comment by lorenzo** [comment]
I would like to merge this... but I think we should investigate if having now 8 file_exists per view file (and element) make any difference in the performance of the framework. Maybe we could offer a shell to cache the locations of each template file?

**Comment by Phillaf** [comment]
I believe php already caches the result of file_exists() calls.

[see this note](http://php.net/manual/en/function.file-exists.php#refsect1-function.file-exists-notes)
> Note: The results of this function are cached. See clearstatcache() for more details.

**Comment by lorenzo** [comment]
@Phillaf that is per request, though

**Comment by Phillaf** [comment]
@lorenzo Oh right! Then I'm not sure how to investigate / benchmark the performance in a reliable way. If we go the caching route, do we even need a shell? Maybe we could always look in the cache first and fallback on the current list of paths? I'll let you guys decide what is worth doing and I can submit a pull request based on your recommendations.

**Comment by markstory** [comment]
The performance delta will really depend on the hosting system. A SSD based host will likely notice no difference. Also the cache used will impact performance deltas. Using file based caches for example, may end up being slower than no caching. I've not done any actual tests, but that is my gut feeling.

**Comment by Phillaf** [comment]
So would the solution be something like this?
- Provide a shell to build the paths in cache
- Provide an opt-in option in the config to turn the feature on

With a shell that runs through all folders, we would cache theme/plugin/controller/actions as keys in a tree-style array with filesystem paths as the values. Then if the config option is set to true we lookup the cache first and if nothing is found we fallback on current paths handling?

**Comment by markstory** [comment]
@Phillaf Before we consider a caching shell, it would be good to have proof that the view lookups could be improved with a persistent cache. I don't think we have concrete proof that it will.

**Pull request closed without merging**